File: Daily Challenge/March 2025/House Robber IV.cs
public class Solution {
  public int MinCapability(int[] nums, int k) {
    int left = 1, right = 1000000000;

    while (left <= right) {
      int mid = (left + right) / 2;
      
      if (CanRob(nums, k, mid)) {
        right = mid - 1;
      } else {
        left = mid + 1;
      }
    }

    return left;
  }

  private bool CanRob(int[] nums, int k, int capability) {
    int houses = 0;
    int cnt = 0;

    for (int i = 0; i < nums.Length; ++i) {
      if (nums[i] <= capability) {
        cnt++;
      } else {
        houses += (cnt + 1) / 2;
        cnt = 0;
      }
    }
    houses += (cnt + 1) / 2;

    return houses >= k;
  }
}

File: Daily Challenge/March 2025/House Robber IV.cpp
class Solution {
public:
  int minCapability(vector<int>& nums, int k) {
    int left = 1, right = 1e9;

    while (left <= right) {
      int mid = (left + right) / 2;
      if (isPossible(nums, k, mid)) {
        right = mid - 1;
      } else {
        left = mid + 1;
      }
    }

    return left;
  }

private:
  bool isPossible(vector<int>& nums, int k, int mid) {
    int count = 0, houses = 0;
    for (int i = 0; i < nums.size(); ++i) {
      if (nums[i] <= mid) {
        count++;
      } else {
        if (count > 0) {
          houses += (count + 1) / 2;
        }
        count = 0;
      }
    }
    if (count > 0) {
      houses += (count + 1) / 2;
    }

    return houses >= k;
  }
};